### ISSUE TITLE
Odd interaction of Generic, Mapping, Hashable

### ISSUE THREAD

**Issue by Dr-Irv** [python/mypy#12506]
**Bug Report**
There seems to be an odd interaction of `Generic`, `Mapping` and `Hashable` as illustrated by the example below.  In the example, `myfun2()` which has a `Union` has an issue in matching the argument, whereas `myfun3()`, which has the same argument list, works fine if you split the `Union` into two overloads.  `myfun4()` and `myfun5()` work as well by substituting `Mapping` with `Dict` and `Hashable` with `Union[int, str]`, respectively
**To Reproduce**

```python
from typing import Mapping, Hashable, Any, Union, Generic, TypeVar, overload, Dict

S1 = TypeVar("S1")


class A(Generic[S1]):
    # fails
    def myfun2(self, m: Union[Mapping[Any, Hashable], Hashable]):
        pass

    # Split the Union into two overloads works
    @overload
    def myfun3(self, m: Mapping[Any, Hashable]):
        ...

    @overload
    def myfun3(self, m: Hashable):
        ...

    def myfun3(self, m: Union[Mapping[Any, Hashable], Hashable]):
        pass

    # Use Dict instead of Mapping works
    def myfun4(self, m: Union[Dict[Any, Hashable], Hashable]):
        pass

    # Use Union[int, str] instead of Hashable works
    def myfun5(self, m: Union[Mapping[Any, Union[int, str]], Union[int, str]]):
        pass


a: A = A()
a.myfun2({1: 1})
a.myfun2("abc")
a.myfun3({1: 1})
a.myfun3("abc")
a.myfun4({1: 1})
a.myfun4("abc")
a.myfun5({1: 1})
a.myfun5("abc")
```

**Expected Behavior**

No error reported.

**Actual Behavior**

```
mapclass.py:23: error: Argument 1 to "myfun2" of "A" has incompatible type "Dict[int, int]"; expected "Hashable"
```

It seems that it dropped the `Mapping[Any, Hashable]` as a potential argument for `myfun2`

If `Mapping` is replaced with `Dict`, there is no problem.  If `Hashable` is replaced, there is no problem.  If you don't have a `Generic` class, there is no problem.

**Your Environment**

- Mypy version used:  0.942
- Mypy command-line flags:  none
- Mypy configuration options from `mypy.ini` (and other config files):  none
- Python version used:  3.9.12
- Operating system and version:  Windows 10

This is related to typing work I'm doing for pandas.

**Comment by AlexWaygood** [comment]
`Hashable` isn't particularly useful for static-typing purposes at the moment, I'm afraid. This is the comment in [typeshed's definition of `Hashable`](https://github.com/python/typeshed/blob/4c9dc43c87bfd1ee80c4243842aae27f0d924a6b/stdlib/typing.pyi#L667-L669):

```
    # TODO: This is special, in that a subclass of a hashable class may not be hashable
    #   (for example, list vs. object). It's not obvious how to represent this. This class
    #   is currently mostly useless for static checking.
```

**Comment by Dr-Irv** [comment]
Does `mypy` use the `typeshed` version of `Hashable`?  Or the version shipped as part of the python standard library, which seems to be different.

Also, the code above works fine with pyright without any report of any issues.

**Comment by AlexWaygood** [comment]
mypy has no knowledge of the standard library; the only thing it sees is the typeshed stubs. If pyright has managed to implement special-casing to handle this case, that's great!

**Comment by JelleZijlstra** [comment]
This doesn't seem like it should need special casing, though. More likely there is an issue with type matching in Unions.

**Comment by Dr-Irv** [comment]
> This doesn't seem like it should need special casing, though. More likely there is an issue with type matching in Unions.

Yes, I think that is the case.

**Comment by erictraut** [comment]
Pyright doesn't have any special casing here. I agree with @JelleZijlstra that this shouldn't require any special casing. The `Hashable` class is a perfectly sound protocol class, and `int` implements that protocol. So this looks like a legit bug in mypy.

**Comment by AlexWaygood** [comment]
I stand corrected!

**Comment by erictraut** [comment]
This bug appears to have been fixed in the 1.0.0 release.

**Comment by Dr-Irv** [comment]
> This bug appears to have been fixed in the 1.0.0 release.

Verified with 1.4.1 that it is fixed.
